NXCM-3986: Improve logs

cstamas opened the pull request:
There is a complex set of conditions when to log what, when to log
stack traces and when not to log stack traces. This would need a fix/cleanup
but for now, this change simply makes to log all HTTP 500 Internal
error responses _with_ stack trace, as previously they were logged
without it, and it did not make any sense and caused information loss.

This seems a long outstanding issue, as original commit implementing
this was unchanged for a long time, and it was causing that all
StorageException (that subclass IOException) gets logged _without_
stack trace, that is bad:

Change introduced in:
fcd4ccf26ad3e963d62e6e45ef06b8428f0d7608

We do have server side logs of the NXCM-3986, but as the cause was
not logged (no stack trace), we lost information why the rename
failed (clearly, it was some sort of FS related problem)

cstamas committed:
NXCM-3986: Improve logs

There is a complex set of conditions when to log what, when to log
stack traces and when not do log stack traces. This would need a fix
but for now, this change simply makes to log all HTTP 500 Internal
error responses _with_ stack trace, as previously they were logged
without it, and it did not make any sense.

This seems as long outstanding issue, as original commit implementing
this was unchanged for a long time, and it was causing that all
StorageException (that subclass IOException) gets logged _without_
stack trace, that is bad:

Change introduced in:
fcd4ccf26ad3e963d62e6e45ef06b8428f0d7608

We do have server side logs of the NXCM-3986, but as the cause was
not logged (no stack trace), we lost information why the rename
failed (clearly, it was some sort of FS related problem)

cstamas committed:
NXCM-3986: Preserve original cause

peterlynch commented:
AbstractResourceStoreContentPlexusResource#handleException - why not fix the javadoc on the way through(params)?
Also I was hoping that method at least had some unit tests, especially now we are changing what it catches

cstamas commented:
We are not changing what it throws, I merely just moved the catch block (literally), and am catching IOEx instead of StorageEx, that is a subclass of StorageEx...

cstamas commented:
ah... catches. Well, `StorageException` was already catch, and if you read whole method. `IOException` was going under the "else" branch. With this change, in case of IOException same happens, as it was before: `throw new ResourceException( Status.SERVER_ERROR_INTERNAL, t );` is executed... hm, this might need a cleanup too.

jdillon commented on nexus/nexus-proxy/src/main/java/org/sonatype/nexus/proxy/storage/local/fs/DefaultFSPeer.java:
```diff
@@ -453,7 +453,7 @@ protected void handleRenameOperation( File hiddenTarget, File target )
 
                 throw new IOException( "Cannot rename file \"" + hiddenTarget.getAbsolutePath() + "\" to \""
-                    + target.getAbsolutePath() + "\"! " + e.getMessage() );
+                    + target.getAbsolutePath() + "\"! " + e.getMessage(), e );
```
String.format() probably a little better here?

```
throw new IOException( String.format("Cannot rename file '%s' to '%s': %s",
    hiddenTarget.getAbsolutePath(), target.getAbsolutePath(), e.getMessage()), e );
```

cstamas replied:
Fixed

peterlynch replied:
-1 
String.format is evil sauce in logging - if you are trying to make it look like what slf4j does, suggest you write a wrapper string formatter that never could throw an exception if you mistyped a token.

jdillon replied:
This isn't logging this is building an exception message.

peterlynch replied:
well duh :) even worse then

jdillon replied:
You are objecting to printf bitching if you gave it %something it didn't comprehend only?  Could either use the slf4j {} thingy or IMO better to use guava-like thingy that only knows about %s, though I don't think that is public we'd have to dup that code.

cstamas committed:
NXCM-3986: Made text handling a bit more concise

cstamas commented:
Rebased against latest master.

cstamas committed:
NXCM-3986: Removed duplicated if branches

Before, there was a separate handling of StorageException's
that was loosing stack trace. Also, superclass of StorageException
is IOException, and it was logged _with_ stack trace (in the else
branch). But both were creating a response of HTTP 500 internal error
(so, ultimately the only difference was the lack of the stack trace
in case of StorageException).

After this change, the StorageExceptions are handled in same way
as their superclass IOException, and will end up in logs _with_
stack trace too.

cstamas commented:
Removed the duplicated logic of StorageException handling. Explained in commit message, but quoting it here too:

Before this fix, there was a separate handling of StorageException's
that was loosing stack trace. Also, superclass of StorageException
is IOException, and it was logged _with_ stack trace (in the else
branch). But both were creating a response of HTTP 500 internal error
(so, ultimately the only difference was the lack of the stack trace
in case of StorageException).

After this change, the StorageExceptions are handled in same way
as their superclass IOException, and will end up in logs _with_
stack trace too and unmodified HTTP 500 response code.

peterlynch commented:
+1 other than direct use of String.format

cstamas commented:
CI scheduled:
https://builds.sonatype.org/view/nexus/job/nexus-oss-its-feature/299/

cstamas commented:
CI voted +1, with one intermittent (and unrelated) failure on Win slave

adreghiciu commented:
+1
